code/datums/abilities/flock/flockmind.dm

```diff
@@ -286,55 +359,116 @@
-	targeted = 0
+	targeted = FALSE
 	cooldown = 0
 
 /datum/targetable/flockmindAbility/controlPanel/cast(atom/target)
 	if(..())
-		return 1
-	var/client/user = holder.owner.client
+		return TRUE
 	var/mob/living/intangible/flock/flockmind/F = holder.owner
-	var/chui/window/flockpanel/panel = F.flock.panel
-	if(isnull(user) || isnull(F) || isnull(panel))
-		return 1
-	panel.Subscribe(user)
+	F.flock.ui_interact(F, F.flock.flockpanel)
 
 ////////////////////////////////
 
 /datum/targetable/flockmindAbility/createStructure
 	name = "Fabricate Structure"
 	desc = "Create a structure tealprint for your drones to construct onto."
 	icon_state = "fabstructure"
-	cooldown = 4 SECONDS
+	cooldown = 0
 	targeted = 0
 
 /datum/targetable/flockmindAbility/createStructure/cast()
-	var/resourcecost = null
-	var/structurewantedtype = null
 	var/turf/T = get_turf(holder.owner)
 	if(!istype(T, /turf/simulated/floor/feather))
 		boutput(holder.owner, "<span class='alert'>You aren't above a flocktile.</span>")//todo maybe make this flock themed?
-		return 1
+		return TRUE
+	if(locate(/obj/flock_structure/ghost) in T)
+		boutput(holder.owner, "<span class='alert'>A tealprint has already been scheduled here!</span>")
+		return TRUE
 	if(locate(/obj/flock_structure) in T)
 		boutput(holder.owner, "<span class='alert'>There is already a flock structure on this flocktile!</span>")
-		return 1
+		return TRUE
+
+	var/list/friendlyNames = list()
+	var/mob/living/intangible/flock/flockmind/F = holder.owner
+	for(var/datum/unlockable_flock_structure/ufs as anything in F.flock.unlockableStructures)
+		if(ufs.check_unlocked())
+			friendlyNames += ufs.friendly_name
+
+
 	//todo: replace with FANCY tgui/chui window with WHEELS and ICONS and stuff!
-	var/structurewanted = tgui_input_list(holder.owner, "Select which structure you would like to create", "Tealprint selection", list("Collector", "Sentinel"))
-	switch(structurewanted)
-		if("Collector")
-			structurewantedtype = /obj/flock_structure/collector
-			resourcecost = 200
-		if("Sentinel")
-			structurewantedtype = /obj/flock_structure/sentinel
-			resourcecost = 300
+
+	var/structurewanted = tgui_input_list(holder.owner, "Select which structure you would like to create", "Tealprint selection", friendlyNames)
+
+	if (!structurewanted)
+		return TRUE
+	var/obj/flock_structure/structurewantedtype = null
+	for(var/datum/unlockable_flock_structure/ufs as anything in F.flock.unlockableStructures)
+		if(ufs.friendly_name == structurewanted)
+			structurewantedtype = ufs.structType
+			break
+
 	if(structurewantedtype)
-		var/mob/living/intangible/flock/F = holder.owner
-		F.createstructure(structurewantedtype, resourcecost)
+		return F.createstructure(structurewantedtype, initial(structurewantedtype.resourcecost))
+
+/////////////////////////////////////////
+
+/datum/targetable/flockmindAbility/ping
+	name = "Ping"
+	desc = "Request attention from other elements of the flock."
+	icon_state = "ping"
+	cooldown = 0.1 SECONDS
```

Comment:
please make this 0.3 seconds or so I have played a lot of DotA 2 and i do not need that in my life anymore

Reply:
https://github.com/stonepillars/goon-flock/pull/318